features/telemetry/protocols/bmp/adapter/src/main/java/org/opennms/netmgt/telemetry/protocols/bmp/adapter/openbmp/BmpIntegrationAdapter.java

```diff
@@ -0,0 +1,724 @@
+                                 final BmpMessageHandler handler) {
+        super(adapterConfig, metricRegistry);
+        this.handler = Objects.requireNonNull(handler);
+    }
+
+    private void handleHeartbeatMessage(final Transport.Message message,
+                                        final Transport.Heartbeat heartbeat,
+                                        final Context context) {
+        final Collector collector = new Collector();
+
+        switch (heartbeat.getMode()) {
+            case STARTED:
+                collector.action = Collector.Action.STARTED;
+                break;
+            case STOPPED:
+                collector.action = Collector.Action.STOPPED;
+                break;
+            case PERIODIC:
+                collector.action = Collector.Action.HEARTBEAT;
+                break;
+            case CHANGE:
+                collector.action = Collector.Action.CHANGE;
+                break;
+        }
+
+        collector.sequence = sequence.getAndIncrement();
+        collector.adminId = context.adminId;
+        collector.hash = context.collectorHashId;
+        collector.routers = Lists.transform(heartbeat.getRoutersList(), BmpAdapterTools::address);
+        collector.timestamp = context.timestamp;
+
+        this.handler.handle(new Message(context.collectorHashId, Type.COLLECTOR, ImmutableList.of(collector)));
+    }
+
+    private void handleInitiationMessage(final Transport.Message message,
+                                         final Transport.InitiationPacket initiation,
+                                         final Context context) {
+        final Router router = new Router();
+        router.action = Router.Action.INIT;
+        router.sequence = sequence.getAndIncrement();
+        router.name = initiation.getSysName();
+//        router.name = InetAddressUtils.str(address(??.getAddress())); // TODO: resolve Ip via DNS?
+        router.hash = context.routerHashId;
+        router.ipAddress = context.sourceAddress;
+        router.description = initiation.getSysDesc();
+        router.termCode = null;
+        router.termReason = null;
+        router.initData = initiation.getMessage();
+        router.termData = null;
+        router.timestamp = context.timestamp;
+        router.bgpId = initiation.hasBgpId() ? BmpAdapterTools.address(initiation.getBgpId()) : null;
+
+        this.handler.handle(new Message(context.collectorHashId, Type.ROUTER, ImmutableList.of(router)));
+    }
+
+    private void handleTerminationMessage(final Transport.Message message,
+                                          final Transport.TerminationPacket termination,
+                                          final Context context) {
+        final Router router = new Router();
+        router.action = Router.Action.TERM;
+        router.sequence = sequence.getAndIncrement();
+        router.name = null;
+//        router.name = InetAddressUtils.str(address(??.getAddress())); // TODO: resolve Ip via DNS?
```

Comment:
TODO: Async DNS resolution. Use address for now and defer to separate JIRA issue?

Reply:
Defered to https://issues.opennms.org/browse/NMS-12569